### PR TITLE
Web - App allows user to open request money link of other user and displays amount page twice

### PR DESCRIPTION
**Pull request by dukenv0307**
### Details

Web - App allows user to open request money link of other user and displays amount page twice
### Fixed Issues

$ https://github.com/Expensify/App/issues/23755
PROPOSAL: https://github.com/Expensify/App/issues/23755#issuecomment-1655489223

### Tests

- Request money with invalid `reportID`
1. Open a chat report
2. Click on the plus icon > Request money
3. Enter a valid amount and continue
4. Copy the URL of the confirm page
5. Paste the URL to composer, edit the `reportID` param to an invalid value, and send the message 
6. Click on the URL that is sent and verify that the not found page appears

- The report that cannot create the money request
1. Open a chat report
2. Click on the plus icon > Request money
3. Copy the URL of the selector page
4. Paste the URL to composer, edit the `reportID` param to a report that cannot create the request money. i.e. the ID of the group chat or announce room, and send the message
5. Click on the URL and verify that the not found page appears

- Request money with global FAB
1. Go to FAB >  Request money
2. Verify the selector page is displayed without loading
3. Open the participant page via deep link without entering an amount
4. Verify that it will reset to the selector page
5. Repeat step 3 with the confirm page, and merchant page and verify that the result is the same

- Test the HOC for some other components
1. Open the report detail page
2. Edit the URL with an invalid reportID
3. Verify that after loading, the not found page appears
4. Repeat the same step with report participant page, report setting page, report share code page
5. Verify that the result is the same
- [x] Verify that no errors appear in the JS console

### QA Steps

- Request money with invalid `reportID`
1. Open a chat report
2. Click on the plus icon > Request money
3. Enter a valid amount and continue
4. Copy the URL of the confirm page
5. Paste the URL to composer, edit the `reportID` param to an invalid value, and send the message 
6. Click on the URL that is sent and verify that the not found page appears

- The report that cannot create the money request
1. Open a chat report
2. Click on the plus icon > Request money
3. Copy the URL of the selector page
4. Paste the URL to composer, edit the `reportID` param to a report that cannot create the request money. i.e. the ID of the group chat or announce room, and send the message
5. Click on the URL and verify that the not found page appears

- Request money with global FAB
1. Go to FAB >  Request money
2. Verify the selector page is displayed without loading
3. Open the participant page via deep link without entering an amount
4. Verify that it will reset to the selector page
5. Repeat step 3 with the confirm page, and merchant page and verify that the result is the same

- Test the HOC for some other components
1. Open the report detail page
2. Edit the URL with an invalid reportID
3. Verify that after loading, the not found page appears
4. Repeat the same step with report participant page, report setting page, report share code page
5. Verify that the result is the same
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

- Request money with invalid `reportID`

https://github.com/Expensify/App/assets/129500732/619d9be4-ee65-4cad-b403-2c4b4378900a

- The report that cannot create the money request

[Screencast from 19-09-2023 15:18:08.webm](https://github.com/Expensify/App/assets/129500732/870e86fc-2a32-4634-8a3d-eb1614a6560f)

- Request money with global FAB

[Screencast from 19-09-2023 15:37:37.webm](https://github.com/Expensify/App/assets/129500732/7ed2d650-1709-4500-bec9-1fbcb7124f81)

- Test the HOC for some other components

[Screencast from 19-09-2023 15:44:24.webm](https://github.com/Expensify/App/assets/129500732/45084002-dd47-405e-b55b-a6b9fc6fbb1b)

</details>

<details>
<summary>Mobile Web - Chrome</summary>
- Request money with invalid `reportID`

https://github.com/Expensify/App/assets/129500732/4f553e07-cdc3-4bc7-ac39-731ba44eb565

- The report that cannot create the money request

https://github.com/Expensify/App/assets/129500732/de1b7d0c-a12f-49d0-beab-d5f55f650170

- Request money with global FAB

https://github.com/Expensify/App/assets/129500732/53b75a73-7cb6-4248-b417-4d482aa565d1

- Test the HOC for some other components

https://github.com/Expensify/App/assets/129500732/8d8adbb6-84c0-4469-b04f-3fe7f922cdd8

</details>

<details>
<summary>Mobile Web - Safari</summary>
- Request money with invalid `reportID`

https://github.com/Expensify/App/assets/129500732/4a5b01b4-e526-443d-ac0e-e8949284e9dc

- The report that cannot create the money request

https://github.com/Expensify/App/assets/129500732/6696709f-0354-4805-81e0-a50de999dc4a

- Request money with global FAB

https://github.com/Expensify/App/assets/129500732/5dc8aada-e316-42e4-984e-7ab3b9a1eda2

- Test the HOC for some other components

https://github.com/Expensify/App/assets/129500732/b5497b35-b1ff-451e-9b97-bfa86f0387fc

</details>

<details>
<summary>Desktop</summary>
- Request money with invalid `reportID`

https://github.com/Expensify/App/assets/129500732/84edaf51-9688-4073-923c-c4bad4989497

- The report that cannot create the money request

https://github.com/Expensify/App/assets/129500732/97ad75de-6907-4cc7-a8b6-0833d5c31b12

- Request money with global FAB

- Test the HOC for some other components

https://github.com/Expensify/App/assets/129500732/ab45c029-eaad-40a2-a050-e49c4e10ce22

</details>

<details>
<summary>iOS</summary>
- Request money with invalid `reportID`

https://github.com/Expensify/App/assets/129500732/820a5e2c-da82-4389-8a59-6dd2ab48513e

- The report that cannot create the money request

https://github.com/Expensify/App/assets/129500732/e3084ec9-843f-44c6-822b-585c6089188c

- Request money with global FAB

https://github.com/Expensify/App/assets/129500732/041e47ff-1b6f-4f7f-a5cb-f2363a3fe3d0

- Test the HOC for some other components

https://github.com/Expensify/App/assets/129500732/7db37afc-70c4-4d25-a9d5-225c63c78adf

</details>

<details>
<summary>Android</summary>
- Request money with invalid `reportID`

https://github.com/Expensify/App/assets/129500732/4a787513-8e07-4d59-a940-b6c2a92e3d8a

- The report that cannot create the money request

[23755-1.webm](https://github.com/Expensify/App/assets/129500732/dea15be4-d7d6-4ac5-b998-3d0183cdd11c)

- Request money with global FAB

https://github.com/Expensify/App/assets/129500732/642d044a-6daa-41c3-8df7-eb6c96ddb69a

- Test the HOC for some other components

https://github.com/Expensify/App/assets/129500732/2bf1656f-9a36-4f07-ba06-94d6da2a2667

</details>

### CODE REVIEW

**Comment by sobitneupane** [comment]
@dukenv0307 Let's add test steps for some other components which use `withReportOrNotFound`. Please add some tests for case of Report Not Found and some for regular flow.

**Comment by sobitneupane** [comment]
@dukenv0307 Yup. Test steps look good.

**Comment by sobitneupane** [comment]
> @sobitneupane I updated all screenshots. For the case 3 in desktop, because this issue https://github.com/Expensify/App/issues/27085 always happens in desktop, I will update this after this issue is merged.

This [issue](https://github.com/Expensify/App/issues/27085) is blocking me from testing. Fortunately, I am assigned to the issue. I will try to get it merged today. After that I will continue the review.

Other than that, @dukenv0307 I'm encountering a 'Not Found' page when attempting to request money from FAB.

**Comment by sobitneupane** [comment]
@dukenv0307 

FAB > Request Money is basically what I did. I might have shared a link of Money Request confirmation page before that.

**Comment by sobitneupane** [comment]
@dukenv0307 As of now, [this PR](https://github.com/Expensify/App/pull/27718) only solves the issue for `/request/new/confirmation` but not for `/request/new/confirmation/123`.

Will the case of `/request/new/confirmation/123` handled by this PR?

**Comment by sourcecodedeveloper** [comment]
> @dukenv0307 As of now, [this PR](https://github.com/Expensify/App/pull/27718) only solves the issue for `/request/new/confirmation` but not for `/request/new/confirmation/123`.
> 
> Will the case of `/request/new/confirmation/123` handled by this PR? As we are aware of the expected output of the case, can we handle it here? It doesn't look like a complex case.

@sobitneupane [PR ](https://github.com/Expensify/App/pull/27718) was made to resolve /request/new/confirmation as reported by @situchan. There is no mention of an invalid reportID (/request/new/confirmation/123).

@dukenv0307 please do not resolve /request/new/confirmation as the I have already put in time and effort to resolve the [issue](https://github.com/Expensify/App/pull/27718)

**Comment by sobitneupane** [comment]
@dukenv0307 [This issue](https://github.com/Expensify/App/pull/26149#issuecomment-1727045480) is still reproducible in my end. Did you get chance to look at it?

https://github.com/Expensify/App/assets/25876548/c61fd33c-5a8e-4b45-a954-074d0c780e2a

**Comment by sobitneupane** [comment]
@dukenv0307 Bit surprising, but it's not working on my end. I'm not sure of the exact reason; maybe some beta feature enabled on my account is causing the crash. Could you try it with a highly trafficked account?

**Comment by sobitneupane** [comment]
[Your latest commit](https://github.com/Expensify/App/pull/26149/commits/b87b9857bcb70b5865280ee401d401e4ab12b4fe) solves the issue.

**Review by sobitneupane — COMMENTED** [review]
### Screenshots/Videos
<details>
<summary>Web</summary>


Screenshot 2023-10-05 at 15 02 00

Screenshot 2023-10-05 at 15 06 46
Screenshot 2023-10-05 at 15 09 40

</details>

<details>
<summary>Mobile Web - Chrome</summary>
Screenshot 2023-10-05 at 15 20 25


Screenshot 2023-10-05 at 15 20 53
Screenshot 2023-10-05 at 15 21 17


</details>

<details>
<summary>Mobile Web - Safari</summary>
Screenshot 2023-10-05 at 15 13 28

Screenshot 2023-10-05 at 15 14 13
Screenshot 2023-10-05 at 15 14 22

</details>

<details>
<summary>Desktop</summary>
Screenshot 2023-10-05 at 15 15 06
Screenshot 2023-10-05 at 15 18 28

Screenshot 2023-10-05 at 15 18 17



</details>

<details>
<summary>iOS</summary>
Screenshot 2023-10-05 at 15 22 43


Screenshot 2023-10-05 at 15 21 44
Screenshot 2023-10-05 at 15 21 52


</details>

<details>
<summary>Android</summary>



</details>

**Comment by sobitneupane** [comment]
I am not 100% confident on whether it is the right way to pass the value up the HOC. But overall code looks good and tests well.

**Review by sobitneupane — APPROVED** [review]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

**Comment by dangrous** [comment]
Hi! It looks like this has expanded just a little since the proposal, do you mind giving me a summary of the new ideas? I think the things I noticed are:
* The new `canCreateRequest` method and using `isAllowedToCreateRequest` in `MoneyRequestSelectorPage` - it sort of feels like this makes the changes to `withReportOrNotFound` more unnecessary? but I am technically out of office so may have vacation brain :)
* The `isArchived` check in `MoneyRequestConfirmPage`
* Removing the useEffect from `NewRequestAmountPage` - is this because it will never be shown due to the canCreateRequest stuff up above?
* Not a question - does `doesRequireReportID` or `requiresReportID` make more sense for our parameter name? `isRequireReportID` feels strange to me.

Nothing is standing out as bad, but I would love some of the context before doing a full review - thank you!

**Comment by dangrous** [comment]
Hi! Sorry, still catching up on stuff. Should be able to look at this shortly.

**Comment by dangrous** [comment]
Okay, so apologies for all the questions. I have a couple more clarifications. Your explanations up there made sense though!

So in the new logic:

* If `isRequireReportID` is true, which is the default, the behavior is exactly the same. So that will be the case for all components except MoneyRequestSelectorPage. Is this correct? Do we need to adjust any of the other Money Request flow pages?
* When `isRequireReportID` is false, if there is a reportID in the route, then it goes through the same logic as if `isRequireReportID` is true. But, it never actually uses the reportID from the route. So wouldn't it always show NotFoundPage?
* When `isRequireReportID` is false, if there is NOT a reportID in the route, then it will just show the Wrapped Component. Which in this case would show the MoneyRequestSelectorPage, which in turn shows the FullPageNotFoundView if there is no report. How does this work if the user IS supposed to see the report? Or is there always props.report except when following an incorrect URL like in the initial issue?
* Should `isReportIdInParam` really be named that, or is it more accurately `isReportIDInRoute` (route v param)

I do prefer `shouldRequireReportID` too.

Tests okay, I'm just still struggling to make sure I understand the code. If you could add some comments in - for me and for future developers - that would be great!

File: src/pages/home/report/withReportOrNotFound.js [hunk 1]
```diff
@@ -44,67 +52,74 @@ export default function (WrappedComponent) {
...
-        const shouldShowNotFoundPage = _.isEmpty(props.report) || !props.report.reportID || !ReportUtils.canAccessReport(props.report, props.policies, props.betas);
-
-        // If the content was shown but it's not anymore that means the report was deleted and we are probably navigating out of this screen.
-        // Return null for this case to avoid rendering FullScreenLoadingIndicator or NotFoundPage when animating transition.
-        if (shouldShowNotFoundPage && contentShown.current) {
-            return null;
+        function WithReportOrNotFound(props) {
+            const contentShown = React.useRef(false);
+
+            const isReportIdInParam = !_.isUndefined(props.route.params.reportID);
```
Comment (dangrous) [inline src/pages/home/report/withReportOrNotFound.js:60]:
```suggestion
            const isReportIdInRoute = !_.isUndefined(props.route.params.reportID);
```

File: src/pages/home/report/withReportOrNotFound.js [hunk 2]
```diff
@@ -44,67 +52,74 @@ export default function (WrappedComponent) {
...
+
+            // If we should require reportID or we have a reportID in the route, we will check the reportID is valid or not
+            if (shouldRequireReportID || isReportIdInParam) {
```
Comment (dangrous) [inline src/pages/home/report/withReportOrNotFound.js:63]:
```suggestion
            if (shouldRequireReportID || isReportIdInRoute) {
```

**Review by dangrous — CHANGES_REQUESTED** [review]
Thanks for the explanation! This looks good - one last variable change to make it a bit clearer and then I think we're good to merge!

**Review by dangrous — APPROVED** [review]
Sweet!

**Pull request merged**